### DISCUSSION TITLE
A way to organize the "sort by"

### DISCUSSION THREAD

**Discussion by Hexality** [ppy/osu#18323] [Ideas]
Hi, I came to suggest a way to organize the "Sort By" line
![image](https://user-images.githubusercontent.com/17398632/169124869-d43f5a5e-fb71-4492-829b-f8a6ba2344f0.png)
through the ini/json file like this
![image](https://user-images.githubusercontent.com/17398632/169128063-b8e4ccac-7271-4ee4-a3fe-fa7658ec4efc.png)

**Comment by frenzibyte** [comment]
Unsure there's much point in such complex layout customisation IMO.

**Reply by Hexality** [reply]
well, it randomly ends up getting messed up with the position changing for no reason.
Because of that, skins like mine ends up getting broken.
![image](https://user-images.githubusercontent.com/17398632/169348954-0e636457-2fd7-4747-88c3-63f6bc64ae82.png)
![image](https://user-images.githubusercontent.com/17398632/169349014-1ca031d4-7dfe-4aa7-ae3a-985f22a8930c.png)

Need to keep re-editing the image all the time to fix, definitely less practical than making it a fixed sequence.

**Reply by bdach** [reply]
how are you even skinning this? i don't think song select supports UI customisation of this kind?

**Reply by Hexality** [reply]
it does :v

**Reply by Hexality** [reply]
you're talking to someone who has been customizing lazer since the day of the `usable skin editor`
![image](https://user-images.githubusercontent.com/17398632/169573163-b4581eb2-e597-48a7-b209-0b1f51eee69a.png)

**Reply by Hexality** [reply]
btw @peppy, somehow the skin name bug is still happening to some extent, haven't this been fixed already?

**Reply by frenzibyte** [reply]
Please refrain from mixing other issues like that. Open a separate discussion thread with details.

**Reply by bdach** [reply]
@hexality let me be *very* clear about one thing: despite your confident (and condescending) claims, there is no *actual* support for skinning song select UI. what you are doing there is hacking around it with sprites in a way that was *not* intended, with text roughly placed to fit the controls underneath. which is why you want to fix the order of items, which is piling hacks on top of hacks.

we will *decidedly* not be doing this to support this sort of hackery, and we may even change skinnable sprites so that clicks don't propagate through them which will break your hack.

**Reply by Hexality** [reply]
Sorry for what I'm going to say know even knowing that you're a collaborator but, what you're saying basically denies everything that has been done to stable until now then, yeah its sprite "hacking", but its the same thing everyone has been doing with osu skins since ages on stable. 
Skinning support is limited on stable and still in early stage on lazer, but it doesn't mean using this type of skinning isn't something that we can't do. The option to d&d sprites into the game and use as part of the skinning system is what we have to use right now, and if it is usable and the only thing available, its not just a hackery, it's a temporary solution for customizability until all features are released.

MY point is that allowing custom positioning would allow users to leave the placement in a way that they can find and use it faster (muscle memory), it's NOT just a feature for skinning, it would just casually help my skinning thing too.

**Reply by integerrr** [reply]
You can do whatever you want with the skin editor as you see fit, but asking the dev team to endorse and conform to your hacky way of "skinning" the song select when there's further planned features to come is actually, completely missing the point.

They are not saying this wont ever be supported but rather not in such a way i.e. through some sort of config file, like suggested below by frenzi, although dont expect to get a nice reply every time if you decide to leave comments that rub people the wrong way

**Reply by Walavouchey** [reply]
The way I see it, the ideal way of skinning UI elements would be to be able to replace drawables directly with sprites and move them around. Placing sprites on top of UI components, like skinning in stable, isn't good for performance reasons and causes problems for lower-end devices. What I mean is that @Hexality yes, that way of skinning isn't optimal and would not be the way skinning will be expected to work in the future. What you can do for now if you want to start skinning with what features are available, is to swap the order of the buttons in the sprite yourself.

> we will decidedly not be doing this to support this sort of hackery, and we may even change skinnable sprites so that clicks don't propagate through them which will break your hack.

I can understand that when skinning support is more developed (perhaps in the way I described it), disabling input propagation through them would make sense, but I would like to point out that for sprites with a lot of transparency (say you want a character in the middle of the screen like in the picture Hexality posted), as well as depending on where you put the sprite, blocking input will only cause UX issues, since you're essentially arbitrarily blocking input on a big region of the screen. Perhaps custom Z-ordering would help alleviate this issue, but it would certainly restrict creativity in skinning.

I don't think any of the points raised here aren't valid, but devs and skinners have different perspectives on this matter and I hope the discussion can continue with a more toned-down temperature.

**Comment by Walavouchey** [comment]
One would hope changing button layouts like this would be possible with the skin editor in the future

**Reply by bdach** [reply]
there is also such a thing as too much customisation

**Reply by Hexality** [reply]
> there is also such a thing as too much customisation

noop, as peppy stated on his blog before
> [...]
> Don’t worry, we are adding support for customisation of the UI layout, to the point you’ll be able to make it look like it used to (or potentially completely different!).
> [...]

**Reply by frenzibyte** [reply]
Sure, once the skin editor actually allows you to customise the layout of song select, rather than adding such setting to the global game config and implement translating it from/to layout positions.
